surfacers/prometheus/prometheus.go

```diff
@@ -104,6 +105,9 @@ type PromSurfacer struct {
 	// corresponding metric string to the provided io.Writer.
 	dataWriter func(w io.Writer, pm *promMetric, dataKey string)
 
+	// A mutex to avoid data access race conditions for EventMetrics.
+	mutex *sync.RWMutex
```

Comment:
I am little torn about using mutex to ensure exclusion between recordMetric and writeData. My thinking was that HTTP handler (which calls "writeData") will be called much fewer times compared to "recordMetric". For example, it's quite unlikely that HTTP handler will be called more than once per second, while calls to recordMetric will scale with the number of targets (for 100 targets, it will be called about 10 times per second with default configuration of exporting stats every 10s).

I can think of couple of alternatives here:

1.  Inside the handler, pass both HTTP response writer and a newly created done channel to the forever running loop and have the handler watch that newly created channel:
```
type httpWriter struct {
  w *http.ResponseWriter
  done chan struct{}
}
```
in handler (line 156):
```
http.HandleFunc(ps.c.GetMetricsUrl(), func(w http.ResponseWriter, r *http.Request) {
  doneCh := make(chan struct{}, 1)
  ps.queryChan <- &httpWriter{w, doneCh}
  <-doneCh
}
```

In the forever loop (line 144):
```
  	case hw := <-ps.queryChan:
                         ps.writeData(hw.w)
			close(hw.doneCh)
         }
```
2. Other alternative is to force the serialization of the HTTP handler call:
```
http.HandleFunc(ps.c.GetMetricsUrl(), func(w http.ResponseWriter, r *http.Request) {
  ps.httpMutex.Lock()
  defer ps.httpMutex.Unlock()

  // Leave the rest same
  ps.queryChan <- w
  <-done
}
```
Enforcing serialization in HTTP handler is not so bad given that we don't expect folks to call /metrics HTTP handler a whole lot (typically a target is scraped by a single prometheus task at 1s-10s frequency).

What do you think about these two approaches.

Reply:
Awesome! I prefer 1 method so will fix this soon! Sorry for late.